### DISCUSSION TITLE
Future of Move Hub support

### DISCUSSION THREAD

**Discussion by laurensvalk** [pybricks/support#218] [General]
This is a very open ended discussion: should we continue to support Pybricks on Move Hub? If so, how would like to see us do it? Please share your ideas.

### Background: limited space

Move Hub does not have much free space for firmware (only 106KiB). We are using more than 95% of this already.

By contrast, the City Hub and the Technic Hub have about 232KiB free, which should be plenty for a while.

This leads to implementation issues and questions like #140.

### Consistency vs. Features

In principle, we could apply a fine-grained approach and enable some but not all new features on Move Hub. But this could get quite complicated. Any thoughts?

### Question: Are you using Move Hub?

How are you using it? Mostly for special cases? Or mostly just like the other hubs? What are your thoughts on the issues above?

Are you happy with the version as it is today? What if we continued supporting it, just without any new features going forward?

### Question: Are you *not* using Move Hub?

Should Move Hub hold back the other hubs by enabling only features that work on all hubs? If yes, why? If not, why not?

**Comment by BertLindeman** [comment]
**Consistency vs. Features**
The movehub is hardware-wise very limited. So limited in functionality is a must. It cannot be helped.
If that is possible with clear error messages that would be preferred. 

- very distant future
Or might there be a possibility to validate a python program that is meant for the movehub?
Just read the movehub program with a python program and then report the discrepancies and/or over-used memory?



**Question: Are you using Move Hub?**
I use the movehub not very often, or I may come across a MOC that uses it.
But most of the time, I use the movehub as pybricks-hub to see what I need to change on a program for pybricks (on another hub)
(Or recovery after too much memory is used)

> Are you happy with the version as it is today? What if we continued supporting it, just without any new features going forward?

For the time being, leave the movehub as is (and supported) and evaluate in "say" a year?

**Question: Are you not using Move Hub?**
> Should Move Hub hold back the other hubs by enabling only features that work on all hubs? If yes, why? If not, why not?

The other hubs should go forward. As stated above the movehub is limited, it cannot be helped.

**wild thought**
In other languages, there is support to ignore source lines between some **#ifdef** / **#endif** statements or similar.
My python knowledge is lacking here.
As an example, I have a program to identify the devices connected to the hub.
This involves many text variables, so a lot of memory. For the movehub, I leave the text out.
Things like that.

**Reply by icefloe01** [reply]
I just purchased a Move Hub as it was the only Hub available on the Lego site that was also mentioned for use with the rollercoaster. However the medium motor was not available so I purchased the large one only to find out it's not compatible through the Boost app. Philo pointed me here because he said Pybricks will work with it.

**Reply by icefloe01** [reply]
I'm surprised the Move is less capable than the others considering the built in motors and how many projects it's supposed to work with. That said, if it's that much of a pain to deal with and if hardly anyone uses it anymore, then yes it's probably a wiser decision to drop it.

**Reply by laurensvalk** [reply]
It’s the first (oldest) hub in this new system, which is probably why.

But you’re using it 😄 (and you’re probably not the only one)

**Reply by icefloe01** [reply]
I chose it over the city hub for two reasons: availability and lack of information. By just looking at it I thought it had to be the most capable hardware. I didn't realize that the Technic/City hub, minus the built in motors, could do the same things better than the Move hub. And it's still out of stock or else I would have already ordered it. I'm guessing it's due to the global chip shortage which has affected my industry as well. I work for a local A/V company that installs Control 4 for home automation. We've had items in backorder for almost a year. I think Lego's biggest issue in my mind is lack of information and they don't really seem to care. Trying to find out from them how this stuff works is like pulling teeth in some cases. And their Boost app doesn't even support modern devices. 🤯

**Comment by laurensvalk** [comment]
A couple of updates/observations since making this post:

- People *are* still using the Move Hub
- We've made good progress with size optimizations
- We've been able to incorporate most new features on Move Hub too.
- Move Hub is going to be supported in V3.2.

**Comment by MakeMagazinDE** [comment]
Despite some lacking features it provides a nice start to build for beginners as you can build so fast a moving platform for experiments.

**Reply by laurensvalk** [reply]
Thank you. Are you building something with Pybricks?

**Reply by MakeMagazinDE** [reply]
I am preparing an article about Pybricks :)

**Reply by laurensvalk** [reply]
Sounds good. Feel free to reach out if you have any questions!

**Comment by antonvh** [comment]
Is the PUPDevice disabled on Move/boost hub because of memory?

**Reply by dlech** [reply]
Yes. :cry:

**Reply by frega** [reply]
Just in case somebody else needs to do this - I was able to get PUPDevice/IODevices to run on the MoveHub. I didn't need the robotics module:

 https://github.com/frega/pybricks-micropython/compare/v3.6.1...3.6.1-movehub-iodevices

**Comment by Leolelion2112** [comment]
Honestly, the Move Hub is a **really** cheap hub. I think it deserves to be maintained as i've seen its price on Bricklink for like 8-20 bucks, and even the full set (Creative Toolbox) for like 45-100 bucks. You might make a good starter bot with 50 bucks if you do it wisely with Studio 2.0 for example.

**Reply by dlech** [reply]
We keep finding ways to reduce the firmware size, so Move hub is safe for now. :smile:
